scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/job/termination/handlers/TerminateReplicateTaskHandler.java

```diff
@@ -0,0 +1,240 @@
+		List<InternalTask> toReplicate = new ArrayList<>();
+
+		// find the tasks that need to be replicated
+		for (InternalTask ti : internalJob.getIHMTasks().values()) {
+			List<InternalTask> tl = ti.getIDependences();
+			if (tl != null) {
+				for (InternalTask ts : tl) {
+					if (ts.getId().equals(initiator.getId()) && !toReplicate.contains(ti)) {
+						// ti needs to be replicated
+						toReplicate.add(ti);
+					}
+				}
+			}
+		}
+
+		// for each initial task to replicate
+		for (InternalTask todup : toReplicate) {
+
+			// determine the target of the replication whether it is a block or
+			// a single task
+			InternalTask target = null;
+
+			// target is a task block start : replication of the block
+			if (todup.getFlowBlock().equals(FlowBlock.START)) {
+				String tg = todup.getMatchingBlock();
+				for (InternalTask t : internalJob.getIHMTasks().values()) {
+					if (tg.equals(t.getName())
+							&& !(t.getStatus().equals(TaskStatus.FINISHED) || t.getStatus().equals(TaskStatus.SKIPPED))
+							&& t.dependsOn(todup)) {
+						target = t;
+						break;
+					}
+				}
+				if (target == null) {
+					logger.error("REPLICATE: could not find matching block '" + tg + "'");
+					continue;
+				}
+			}
+			// target is not a block : replication of the task
+			else {
+				target = todup;
+			}
+
+			// for each number of parallel run
+			for (int i = 1; i < runs; i++) {
+
+				// accumulates the tasks between the initiator and the target
+				Map<TaskId, InternalTask> dup = new HashMap<>();
+				// replicate the tasks between the initiator and the target
+				try {
+					target.replicateTree(dup, todup.getId(), false, initiator.getReplicationIndex() * runs, 0);
+
+				} catch (Exception e) {
+					logger.error("REPLICATE: could not replicate tree", e);
+					break;
+				}
+
+				((JobInfoImpl) internalJob.getJobInfo()).setNumberOfPendingTasks(
+						((JobInfoImpl) internalJob.getJobInfo()).getNumberOfPendingTasks() + dup.size());
+
+				// pointers to the new replicated tasks corresponding the begin
+				// and
+				// the end of the block ; can be the same
+				InternalTask newTarget = null;
+				InternalTask newEnd = null;
+
+				// configure the new tasks
+				for (Entry<TaskId, InternalTask> it : dup.entrySet()) {
+					InternalTask nt = it.getValue();
+					nt.setJobInfo(((JobInfoImpl) internalJob.getJobInfo()));
+					int dupIndex = getNextReplicationIndex(InternalTask.getInitialName(nt.getName()),
+							nt.getIterationIndex());
+					internalJob.addTask(nt);
+					nt.setReplicationIndex(dupIndex);
+					assignReplicationTag(nt, initiator, false, action);
+				}
+				changesInfo.newTasksAdded(dup.values());
+
+				// find the beginning and the ending of the replicated block
+				for (Entry<TaskId, InternalTask> it : dup.entrySet()) {
+					InternalTask nt = it.getValue();
+
+					// connect the first task of the replicated block to the
+					// initiator
+					if (todup.getId().equals(it.getKey())) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS134)